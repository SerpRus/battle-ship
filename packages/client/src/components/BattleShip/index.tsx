import React, { useRef, useEffect, useState } from 'react';
import StartStep from '../../widgets/game/steps/start-step';
import GameStep from '../../widgets/game/steps/game-step';
import EndStep from '../../widgets/game/steps/end-step';
import { ShipsType } from '../../widgets/game/utils/generate-ships';
import ships from '../../widgets/game/elements/ships';
import { PLAYER_BOARD_POSITION } from '../../widgets/game/utils/constants';

type GameStepsType = {
  start: typeof StartStep;
  game: typeof GameStep;
  end: typeof EndStep;
};

export default function BattleShip() {
  const canvas = useRef<null | HTMLCanvasElement>(null);

  const [gameStep, setGameStep] = useState('start');
  const [playerShips, setPlayerShips] = useState([]);

  const clickRef = useRef<null | ((e: React.MouseEvent<HTMLElement>) => void)>(
    null
  );

  useEffect(() => {
    const gameSteps: GameStepsType = {
      start: StartStep,
      game: GameStep,
      end: EndStep,
    };

    const canvasElement = canvas.current;
    if (!canvasElement) {
      return;
    }

    const ctx = canvasElement.getContext('2d');

    if (!ctx) {
      return;
    }

    ctx.clearRect(0, 0, canvasElement.width, canvasElement.height);

    const currentStep = new gameSteps[gameStep as keyof GameStepsType](
      ctx,
      canvasElement,
      setGameStep,
      setPlayerShips as React.Dispatch<React.SetStateAction<ShipsType>>,
      playerShips
    );

    currentStep.render();

    ships(
      ctx,
      {
        x: PLAYER_BOARD_POSITION.x,
        y: PLAYER_BOARD_POSITION.y,
      },
      playerShips
    );

    if (currentStep.clickHandler) {
      clickRef.current = currentStep.clickHandler;
    }
  }, [gameStep, playerShips]);

  return (
    <canvas
      ref={canvas}
      width={1041}
      height={601}
      onClick={e => {
        if (clickRef.current) {
          clickRef.current(e);
        }
      }}
    />
  );
}
