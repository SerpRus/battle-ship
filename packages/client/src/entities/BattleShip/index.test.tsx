import * as renderer from 'react-test-renderer';
import 'jest-canvas-mock';
import React from 'react';
import { shallow, configure, ShallowWrapper } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import BattleShip from './index';

type TObjeect = Record<string, unknown>;

configure({ adapter: new Adapter() });

describe('Game', () => {
  let wrapper: ShallowWrapper<
    any,
    Readonly<TObjeect>,
    React.Component<TObjeect, TObjeect, any>
  >;

  const canvas: HTMLCanvasElement = document.createElement('canvas');
  const clickRef: {
    current: null | ((e: React.MouseEvent<HTMLElement>) => void);
  } = { current: null };
  canvas.width = 1041;
  canvas.height = 601;
  canvas.onclick = e => {
    if (clickRef.current) {
      clickRef.current(e as unknown as React.MouseEvent<HTMLElement>);
    }
  };

  beforeEach(() => {
    wrapper = shallow(<BattleShip />);
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<BattleShip />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('width of canvas element is correctly', () => {
    const bool = wrapper.find(HTMLCanvasElement).props().width === 1041;
    wrapper.update();
    expect(bool).toBe(true);
  });
});
