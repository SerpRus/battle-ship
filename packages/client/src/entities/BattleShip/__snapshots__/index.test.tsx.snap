// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders correctly 1`] = `
<canvas
  height={601}
  onClick={[Function]}
  width={1041}
/>
`;
